[core] New report format html-report-v2.xslt to provide html with datatable and chart features

mohan-chinnappan-n opened the pull request:
## Describe the PR

Added pmd-report-v2.xslt to provide html with
-  datatable (search, column sorting and pagination)
-  chart features

- Screenshot
<img width="1622" alt="Screen Shot 2022-09-12 at 7 17 02 PM" src="https://user-images.githubusercontent.com/5963194/189774640-4e714281-91f8-4d6d-9cb7-1b72df70c5a4.png">

Also updated CLI help with a line for this xslt

## Related issues
 

- Fixes #

## Ready?

<!-- If you feel like you can help to check off the following tasks, that'd be great. If not, don't worry - we will take care of it. -->

- [ ] Added unit tests for fixed bug/feature
- [ ] Passing all unit tests
- [ ] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [ ] Added (in-code) documentation (if needed)

pmd-test commented:
<!--
  0 Errors
  0 Warnings
  1 Message: Compared to master:<br />
This...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 0 violations,<br />
introduces 0 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/c260c241ab41445ba26a54e56a95bbb3/diff1/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

adangel commented on pmd-core/etc/xslt/pmd-report-v2.xslt:
```diff
@@ -0,0 +1,251 @@
+<?xml version="1.0"?>
+<xsl:stylesheet version="1.0" xmlns:xsl="http://www.w3.org/1999/XSL/Transform"
```
Since we produce html, it's probably better to explicitly configure html output:

```xml
<xsl:output method="html"/>
```

Otherwise the default might be xml.

adangel commented on pmd-core/etc/xslt/pmd-report-v2.xslt:
```diff
@@ -0,0 +1,251 @@
+  <html>
+   <head>
+   <title>PMD Report </title>
+<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@5.2.0/dist/css/bootstrap.min.css" integrity="sha384-gH2yIJqKdNHPEq0n4Mqa/HGKIhSkIHeL5AyhkYV8i59U5AR6csBvApHHNl/vI1Bx" crossorigin="anonymous"/>
```
How should we deal with updates of these libraries? E.g. bootstrap 5.2.1 is now available.
I think, whatever version we use here, it will be outdated very soon. Is this a problem we need to address?

adangel commented on pmd-core/etc/xslt/pmd-report-v2.xslt:
```diff
@@ -0,0 +1,251 @@
+
+  <html>
+   <head>
+   <title>PMD Report </title>
```
Nit: the indentation is not consistent - some lines are indented correctly, some are not...

adangel commented on pmd-core/etc/xslt/pmd-report-v2.xslt:
```diff
@@ -0,0 +1,251 @@
+
+    <xsl:variable name="ruleset" select="@ruleset" />
+    <xsl:choose>
+      <xsl:when test="$ruleset = 'Best Practices'">
```
Note: This mapping might not work, when custom rulesets are used.
Consider to add an "other" counter for all rulesets you couldn't map.

adangel commented on pmd-core/etc/xslt/pmd-report-v2.xslt:
```diff
@@ -0,0 +1,251 @@
+    </xsl:choose>
+          <td><xsl:value-of select="@ruleset"/></td>
+
+          <td> <a href="{@externalInfoUrl}">Details</a></td>
```
Nice, I didn't know this syntax. I assume, the XSL processor will replace this expression with the corresponding value.

adangel commented on pmd-core/etc/xslt/pmd-report-v2.xslt:
```diff
@@ -0,0 +1,251 @@
+      <xsl:variable name="filename" select="@name" />
+      <xsl:variable name="slno" select="position()" />
+         <xsl:for-each select="scan:violation">
+          <xsl:variable name="pos" select="position()" />
+           <xsl:variable name="currPos" select="position()" />
```
These two variables don't seem to be used

adangel commented on pmd-core/etc/xslt/pmd-report-v2.xslt:
```diff
@@ -0,0 +1,251 @@
+      </thead>
+      <xsl:for-each select="scan:file">
+      <xsl:variable name="filename" select="@name" />
+      <xsl:variable name="slno" select="position()" />
```
Please give this variable a better name - I don't know what "slno" is. Maybe "fileNumber"?

adangel reviewed (commented):
Thanks for the PR!

The screenshot looks impressive, especially the charts are a good idea.

For the available report formats, we have them all mentioned in the documentation (see https://pmd.github.io/latest/pmd_userdocs_report_formats.html). Please add your new format there as well (in this PR).

The name could be simply "html-report-v2.xslt" - since it should be clear that this report is for PMD anyway...

In the documentation we can also describe, that using this report format, various javascript/css sources are loaded from the internets... depending on the environment this might not be allowed and/or possible. So, this report format won't work offline.

There we can also describe the additional features (like pdf export, excel export). Although I'm undecided, whether to include all these features, as it smells like doing too much and a very indirect solution (in order to create a excel file, you first need to create a html report...).

Maybe you can spend some time and look at #4132 (which I've just created) and share your thoughts. It would be interesting to know the different use cases (e.g. why we need a PDF export).

adangel commented on pmd-core/etc/xslt/pmd-report-v2.xslt:
```diff
@@ -0,0 +1,251 @@
+         <script> total++; </script>
+    <xsl:choose>
+      <xsl:when test="$priority = 1">
+        <script> p1++; </script>
```
One more thing: If we have a pmd report with many violations, then we would add these script-tags for each violation. This unnecessarily blows up the html output. It should be possible to use jquery to calculate the statistics afterwards from the table data in the document. That's by the way how the datatable gets hold of the data for paging etc. - we can do the same thing.

mohan-chinnappan-n commented:
Hi Andreas,
Thanks for the feedback. Let me work on them and keep you updated...


On Sat, Sep 24, 2022 at 9:13 AM Andreas Dangel ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In pmd-core/etc/xslt/pmd-report-v2.xslt
> <https://github.com/pmd/pmd/pull/4125#discussion_r979242782>:
>
> > +      <xsl:variable name="filename" ***@***.***" />
> +      <xsl:variable name="slno" select="position()" />
> +         <xsl:for-each select="scan:violation">
> +          <xsl:variable name="pos" select="position()" />
> +           <xsl:variable name="currPos" select="position()" />
> +          <tr>
> +          <td><xsl:value-of select="$slno"/></td>
> +          <td><xsl:value-of select="position()"/></td>
> +          <td><xsl:value-of select="$filename"/> </td>
> +          <td><xsl:value-of ***@***.***"/></td>
> +          <td><xsl:value-of ***@***.***"/></td>
> +          <xsl:variable name="priority" ***@***.***" />
> +         <script> total++; </script>
> +    <xsl:choose>
> +      <xsl:when test="$priority = 1">
> +        <script> p1++; </script>
>
> One more thing: If we have a pmd report with many violations, then we
> would add these script-tags for each violation. This unnecessarily blows up
> the html output. It should be possible to use jquery to calculate the
> statistics afterwards from the table data in the document. That's by the
> way how the datatable gets hold of the data for paging etc. - we can do the
> same thing.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pmd/pmd/pull/4125#pullrequestreview-1119267507>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABNP3OVH5LTFOPQCHGAKDWLV735BBANCNFSM6AAAAAAQK5JQO4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

oowekyala commented:
Hi @mohan-chinnappan-n, how is it going with this? Do you think we could try schedule it for the next release (in a month)? Please let us know if you need any help
Cheers

mohan-chinnappan-n commented:
Sure. We can schedule it for the next release..

On Tue, Oct 25, 2022 at 3:13 PM Clément Fournier ***@***.***>
wrote:

> Hi @mohan-chinnappan-n <https://github.com/mohan-chinnappan-n>, how is it
> going with this? Do you think we could try schedule it for the next release
> (in a month)? Please let us know if you need any help
> Cheers
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pmd/pmd/pull/4125#issuecomment-1291021743>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABNP3OWE3XLVES7GTL54O2TWFAWODANCNFSM6AAAAAAQK5JQO4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

mohan-chinnappan-n commented:
Hi Clement,
Planning to finish this by Nov-06-2022.
you have a slack channel, we can communicate.
I have a few ideas like adding new features to PMD.... we can discuss



On Tue, Oct 25, 2022 at 3:20 PM Mohan Chinnappan <
***@***.***> wrote:

> Sure. We can schedule it for the next release..
>
> On Tue, Oct 25, 2022 at 3:13 PM Clément Fournier ***@***.***>
> wrote:
>
>> Hi @mohan-chinnappan-n <https://github.com/mohan-chinnappan-n>, how is
>> it going with this? Do you think we could try schedule it for the next
>> release (in a month)? Please let us know if you need any help
>> Cheers
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/pmd/pmd/pull/4125#issuecomment-1291021743>, or
>> unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ABNP3OWE3XLVES7GTL54O2TWFAWODANCNFSM6AAAAAAQK5JQO4>
>> .
>> You are receiving this because you were mentioned.Message ID:
>> ***@***.***>
>>
>

adangel commented:
Hi @mohan-chinnappan-n , any news on this? Can we help?

mohan-chinnappan-n commented:
Created new html-report-v2.xslt based on the comments and feedback

mohan-chinnappan-n commented:
Hi Andreas,
- Changed the name to pmd-core/etc/xslt/html-report-v2.xslt
- Updated with your comments/feedback.
Hope you can merge it.
I will update the docs as well as after you merge it.

On Fri, Nov 18, 2022 at 9:45 AM Andreas Dangel ***@***.***>
wrote:

> Hi @mohan-chinnappan-n <https://github.com/mohan-chinnappan-n> , any news
> on this? Can we help?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pmd/pmd/pull/4125#issuecomment-1320089501>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABNP3OWL5WSVSDHYDGJGSFDWI6JANANCNFSM6AAAAAAQK5JQO4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

adangel commented on pmd-core/etc/xslt/html-report-v2.xslt:
```diff
@@ -0,0 +1,232 @@
+      {"rule": "Documentation", "count": doc},
+
+      {"rule": "Documentation", "count": ep},
+      {"rule": "Error Prone", "count": perf},
```
```suggestion
      {"rule": "Performance", "count": perf},
```

adangel commented on pmd-core/src/main/java/net/sourceforge/pmd/cli/PMDCommandLineInterface.java:
```diff
@@ -138,11 +138,13 @@ private static String getUnixExample() {
         return "For example on *nix: " + PMD.EOL
                 + launchCmd + " -dir /home/workspace/src/main/java/code -f html -rulesets rulesets/java/quickstart.xml,category/java/codestyle.xml" + PMD.EOL
                 + launchCmd + " -d ./src/main/java/code -R rulesets/java/quickstart.xml -f xslt -property xsltFilename=my-own.xsl" + PMD.EOL
+                + launchCmd + " -d ./src/main/java/code -R rulesets/java/quickstart.xml -f xslt -property xsltFilename=pmd-report-v2.xslt" + PMD.EOL
+                + " - pmd-report-v2.xslt is at https://github.com/pmd/pmd/tree/master/pmd-core/etc/xslt/pmd-report-v2.xslt"
```
```suggestion
                + launchCmd + " -d ./src/main/java/code -R rulesets/java/quickstart.xml -f xslt -property xsltFilename=html-report-v2.xslt" + PMD.EOL
                + " - html-report-v2.xslt is at https://github.com/pmd/pmd/tree/master/pmd-core/etc/xslt/html-report-v2.xslt"
```

adangel reviewed (changes requested):
Thanks for the update!

- Please remove the file "pmd-report-v2.xslt", as you have renamed it. We only need "html-report-v2.xslt".
- Also please add the documentation directly in this PR. This increases the probability that we have the documentation ready when we merge this PR.

adangel reviewed (approved):
This is ready now
